### DISCUSSION TITLE
Nuxt 3 linting?

### DISCUSSION THREAD

**Discussion by ricardofiorani** [nuxt/nuxt#15889] [Questions]
Would it be possible to provide the [linting instructions page](https://typescript.nuxtjs.org/guide/lint/) but for Nuxt 3?
Since the bridge installation suggests to [remove](https://v3.nuxtjs.org/getting-started/bridge#remove-incompatible-and-obsolete-modules) some libs that provided linting rules, right now I have a project being migrated to Nuxt 3 that is unlintable. I tried myself adding the `plugin:vue/vue3-recommended` rules and other similar rules to my `.estlinrc` file but either they work only on `.js` files or don't work well with typescript at all.

Thank you for maintaining and pushing Nuxt forward.

**Comment by franzus5** [comment]
@ricardofiorani works for me

```bash
[yarn add/npm install] -D eslint eslint-plugin-vue typescript @typescript-eslint/eslint-plugin @typescript-eslint/parser vite-plugin-eslint @nuxtjs/eslint-config-typescript
```

`.eslintrc.json`
```json
{
    "env": {
        "browser": true,
        "es2021": true,
        "node": true
    },
    "extends": [
        "eslint:recommended",
        "plugin:vue/essential",
        "plugin:@typescript-eslint/recommended",
        "@nuxtjs/eslint-config-typescript"
    ],
    "parserOptions": {
        "ecmaVersion": "latest",
        "parser": "@typescript-eslint/parser",
        "sourceType": "module"
    },
    "plugins": [
        "vue",
        "@typescript-eslint"
    ],
    "rules": {}
}
```

Add to `nuxt.config.ts`
```typescript
vite: {
    plugins: [
      eslintPlugin()
    ]
}
```

**Reply by yungParrot** [reply]
Thanks @franzus5 for Your config! I can confirm it works
although my `nuxt.config.ts` needed `import eslintPlugin from "vite-plugin-eslint";` in `nuxt.config.ts`
```typescript
import { defineNuxtConfig } from "nuxt3";
import eslintPlugin from "vite-plugin-eslint";

// https://v3.nuxtjs.org/docs/directory-structure/nuxt.config
export default defineNuxtConfig({
  vite: {
    plugins: [eslintPlugin()]
  }
});
```

---

My current `.eslintrc.js`:
```javascript
module.exports = {
  env: {
    browser: true,
    es2021: true,
    node: true
  },
  extends: [
    "eslint:recommended",
    "plugin:vue/essential",
    "plugin:@typescript-eslint/recommended",
    "@nuxtjs/eslint-config-typescript"
  ],
  parserOptions: {
    ecmaVersion: "latest",
    parser: "@typescript-eslint/parser",
    sourceType: "module"
  },
  plugins: ["vue", "@typescript-eslint"],
  rules: {
    quotes: ["error", "double"],
    semi: ["error", "always"],
    indent: ["error", 2],
    "no-multi-spaces": ["error"],
    "no-unused-vars": [
      "error",
      { vars: "all", args: "after-used", ignoreRestSiblings: false }
    ]
  }
};

```
Remember to have ESLint installed and enabled in VSCode.  
[This article](https://miyauchi.dev/posts/vite-vue3-typescript/) was also quite helpful...

**Reply by bodasooqa** [reply]
Cool, thanks for this solution!
Did you manage to run the linter together with `nuxi dev`? So that the code is checked in real time

**Reply by yungParrot** [reply]
@bodasooqa I haven't been working on that project in the last month so I'm not 100% sure...  
I think the linting was working correctly with `nuxi dev`, although I would have to check it later 🤔

**Reply by lucasctd** [reply]
Are you using
 ```html
<script lang="ts" setup>
``` 
?

It keeps throwing warnings saying my imports, functions and refs are not being used.

**Reply by yungParrot** [reply]
@lucasctd 
```vue
<script setup lang="ts">
...
</script>
```

seems to be working in my project and yes, I do use the Composition API with `script setup`. I haven't tested with the Options API :confused: you would have to show me what errors you're getting and what is inside your code...  

@bodasooqa about
> Did you manage to run the linter together with nuxi dev? So that the code is checked in real time

`nuxi dev` doesn't seem to show any linting errors when running the code -> I get linting errors in VSCode by using the ESLint plugin but the output from the CLI doesn't seem to notice any ESLint errors :cowboy_hat_face:

**Reply by RihanArfan** [reply]
I'd recommend `plugin:vue/vue3-recommended` instead of `plugin:vue/essential` - https://eslint.vuejs.org/user-guide/#configuration

**Reply by holyris** [reply]
I can't manage to make it work correctly with Prettier vscode extension, any advices ?

**Reply by Ygoro** [reply]
> Are you using
> 
> ```
> <script lang="ts" setup>
> ```
> 
> ?
> 
> It keeps throwing warnings saying my imports, functions and refs are not being used.

Adding the `'vue/script-setup-no-uses-vars': 'off'` rule, as proposed [here](https://itnext.io/nuxt-3-first-steps-c23d142405c4#:~:text=Right%2C%20rules%20and,Vue%2Drelated%20ones.), resolved this issue for me. Hope this helps ☺️

**Comment by Jellfedora** [comment]
Do you have an eslint config without typescript? I have errors for automatic imports

**Reply by D-Mx** [reply]
Just remove everything related to typescript and replace `@nuxtjs/eslint-config-typescript` with `@nuxtjs/eslint-config`. Although the Nuxt TypeScript ESLint config includes the other one.

Extend the config
```Json
{
  "extends": [
    "@nuxtjs"
  ]
}
```

**Comment by sepiaflux** [comment]
I keep having trouble with "vue/no-v-model-argument" when i use v-model:XYZ .
I also cannot use defineEmits like this:
```
const emit = defineEmits<{
  (e: 'update:open', id: boolean): void
}>()

```
or I get "Unexpected whitespace between function name and paren  func-call-spacing" .
Does anyone have any ideas? My config is as below and I use vite-plugin-eslint as shown above.
```
{
    "env": {
        "browser": true,
        "es2021": true,
        "node": true
    },
    "extends": [
        "eslint:recommended",
        "plugin:@typescript-eslint/recommended",
        "@nuxtjs/eslint-config-typescript",
        "plugin:vue/vue3-recommended"
    ],
    "parserOptions": {
        "ecmaVersion": "latest",
        "parser": "@typescript-eslint/parser",
        "sourceType": "module"
    },
    "plugins": [
        "@typescript-eslint"
    ],
    "ignorePatterns": ["**/generated/graphql.ts"],
    "rules": {
        "vue/multi-word-component-names": "off",
        "vue/no-setup-props-destructure": "off",
        "no-console": "off",
        "vue/no-v-model-argument": "off"
    }
}
```

**Comment by thisismydesign** [comment]
https://stackoverflow.com/a/73718358/2771889

A simple ESLint + Prettier + TypeScript + Nuxt 3 (or Bridge) setup would look like this:

`yarn add --dev eslint prettier eslint-config-prettier eslint-plugin-prettier @nuxtjs/eslint-config-typescript`

`.eslintrc.js`
```js
module.exports = {
  root: true,
  extends: ['@nuxtjs/eslint-config-typescript', 'plugin:prettier/recommended'],
}
```

**Comment by maxdzin** [comment]
I also tried to find a proper way and it seems to work well with this config:
```json
{
  "root": true,
  "env": {
    "browser": true,
    "es2022": true,
    "node": true
  },
  "extends": [
    "@nuxtjs/eslint-config-typescript",
    "plugin:nuxt/recommended",
    "plugin:prettier/recommended"
  ],
  "parserOptions": {
    "ecmaVersion": "latest",
    "parser": "@typescript-eslint/parser",
    "sourceType": "module"
  },
  "plugins": ["@typescript-eslint", "vue", "nuxt"],
  "rules": {}
}
```

**Reply by RihanArfan** [reply]
The `@nuxtjs/eslint-config-typescript` package handles most of what you've specified here. Here's a cleaner config which does the same thing.

```json
{
  "root": true,
  "env": {
    "browser": true,
    "es2022": true,
    "node": true
  },
  "extends": [
    "@nuxtjs/eslint-config-typescript",
    "plugin:prettier/recommended"
  ],
  "parserOptions": {
    "ecmaVersion": "latest",
    "sourceType": "module"
  },
  "rules": {}
}
```

These packages are needed. `yarn add -D eslint eslint-config-prettier @nuxtjs/eslint-config-typescript prettier`

**Comment by steklopod** [comment]
Do we also need ` npm install eslint-plugin-nuxt@latest --save-dev`?

Without this dependency it does not work on my side.

**Reply by maxdzin** [reply]
> Do we also need ` npm install eslint-plugin-nuxt@latest --save-dev`?
> 
> Without this dependency it does not work on my side.

Yes. Same for me, so, it needs to be installed.

**Comment by mrguamos** [comment]
I'm using [antfu's eslint config](https://github.com/antfu/eslint-config)
try it very convenient

**Reply by vin-ni** [reply]
This was the only one working for me.

**Reply by sh-ravan** [reply]
I ended up using this as well. Did have to install `vue-tsc` and `typescript` as devDependencies to get it fully working tho

**Reply by HollyCreep** [reply]
@sh-ravan could you please share your configs ?

**Reply by vin-ni** [reply]
as written under antfu's page :)

```
{
  "extends": "@antfu"
}
```

**Reply by HollyCreep** [reply]
I was getting some TS erros, on imports and modules. But I isntalled it again and somehow, now its working fine. Anyway, thanks for your help !

**Comment by DesKevinMendez** [comment]
I have these configurations.

First: install all dependencies
`yarn/npm -D @typescript-eslint/eslint-plugin @nuxtjs/eslint-config-typescript vite-plugin-eslint eslint`

Add the following config into nuxt.config
```
import eslintPlugin from 'vite-plugin-eslint'

....
  vite: {
    plugins: [
      eslintPlugin({
        emitWarning: true,
        emitError: true
      })
    ]
  }
```

and my .eslintrc.json
```
{
  "env": {
    "browser": true,
    "es2021": true,
    "node": true
  },
  "extends": [
    "eslint:recommended",
    "plugin:vue/essential",
    "plugin:@typescript-eslint/recommended",
    "@nuxtjs/eslint-config-typescript"
  ],
  "parserOptions": {
    "ecmaVersion": "latest",
    "parser": "@typescript-eslint/parser",
    "sourceType": "module"
  },
  "plugins": ["vue", "@typescript-eslint"],
  "rules": {
  }
}

```

and my eslintignore
```
/node_modules
/dist
```

And remember to add the script into package.json

`
    "lint": "eslint --ext .js,.ts,.vue --ignore-path .eslintignore ."`

**Comment by bagaskarala** [comment]
install these dev dependencies:
`eslint typescript @nuxtjs/eslint-config-typescript eslint-config-prettier eslint-plugin-nuxt eslint-plugin-prettier prettier`

---
`.eslintrc.js` - this config below already contained with good vue-nuxt-typescript support
```javascript
module.exports = {
  root: true,
  env: {
    browser: true,
    es2021: true,
    node: true,
  },
  extends: [
    '@nuxtjs/eslint-config-typescript',
    'plugin:nuxt/recommended',
    'plugin:prettier/recommended',
  ],
  parserOptions: {
    ecmaVersion: 'latest',
  },
  rules: {},
};
```

---
`.prettierrc`
```
{
  "semi": false,
  "singleQuote": true
// rest of your style
}
```

**Comment by ndragun92** [comment]
I can also provide what I have in mine

```
// .eslintrc.js
module.exports = {
  env: {
    browser: true,
    es2021: true,
    node: true,
  },
  extends: [
    "plugin:@typescript-eslint/recommended",
    "plugin:nuxt/recommended",
    "plugin:vue/vue3-recommended",
    "plugin:prettier/recommended",
  ],
  parserOptions: {
    ecmaVersion: "latest",
    parser: "@typescript-eslint/parser",
    sourceType: "module",
  },
  plugins: ["@typescript-eslint"],
  rules: {
    "@typescript-eslint/no-explicit-any": "off",
    "@typescript-eslint/no-non-null-assertion": "off",
    "vue/no-v-html": "off",
    camelcase: "off",
    "prettier/prettier": [
      "error",
      {
        endOfLine: "auto",
      },
    ],
  },
};
```

```
// nuxt.config.ts
import eslintPlugin from "vite-plugin-eslint";

export default defineNuxtConfig({
  vite: {
    plugins: [
      eslintPlugin(),
    ],
});

```

```
// package.json
{
  "devDependencies": {
    "@typescript-eslint/eslint-plugin": "^5.46.0",
    "@typescript-eslint/parser": "^5.46.0",
    "eslint": "^8.29.0",
    "eslint-config-prettier": "^8.5.0",
    "eslint-plugin-nuxt": "^4.0.0",
    "eslint-plugin-prettier": "^4.2.1",
    "eslint-plugin-vue": "^9.7.0",
    "prettier": "^2.8.1",
    "vite-plugin-eslint": "^1.8.1",
  }
}
```
